File: 0078-subsets/solution.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> ans = new ArrayList<>();

        generateSubsets(0, nums, res, ans);
        return res;
    }

    public void generateSubsets(int start, int[] nums, List<List<Integer>> res, List<Integer> ans) {
        int n = nums.length;
        // removed ans.size() == n, as subsets can be of any size, not only of n size
        // it can be [], [1,2] etc
        res.add(new ArrayList<>(ans));

        for(int i=start; i<n; i++) {
            ans.add(nums[i]);
            // i+1 , controls next recursion pointer
            generateSubsets(i+1, nums, res, ans);
            ans.remove(ans.size()-1);
        }
    }
    
}

File: 0078-subsets/solution.swift
class Solution {
    func subsets(_ nums: [Int]) -> [[Int]] {
        var n = nums.count
        var res = [Int]()
        var ans = [[Int]]()

        solve(0)

        func solve(_ index: Int) {
            // You’re only adding res to ans when index == n
            // But with the for-loop subset approach, every current res is a valid subset.
            // so commenting index == n
            // if index == n {
                ans.append(res)
                // return
            // }
            // at every number
            // take / skip
            // for loop does going to every number
            // choose 1, choose 2 
            for i in index..<nums.count {
                // choose
                res.append(nums[i])

                // explore choices after i
                solve(i+1)

                // unchoose
                res.removeLast()
            }
            // skip
            // solve(index+1)
        }
        return ans
    }
}
